### ISSUE TITLE
False warning: "unused logical operation that must be used"

### ISSUE THREAD

**Issue by spacekookie** [rust-lang/rust#69466]
In a loop, in some cases, rustc seems to not understand that a boolean expression was used for sideeffects, and warns about unused logical operation.  The following code can quickly replicate the problem.

```rust
use std::sync::{atomic::{AtomicBool, Ordering}, Arc};


fn main() {
    let running = Arc::new(AtomicBool::from(true));
    
    {
        let running = Arc::clone(&running);
        std::thread::spawn(move || {
            running.swap(false, Ordering::Relaxed);
        });
    }
    
    loop {
        // ...
        println!("Running....");
        
        !running.load(Ordering::Relaxed) && break;
    }
    
    println!("Exhausted >.>");
}

```

The expected result is that rustc doesn't output a warning in this case.

### Meta

* rustc version: 1.41.0

**Comment by KrishnaSannasi** [comment]
I don't think this is the intended use of logical operators, you should use an `if`, when side effects matter. This lint should help people stay away from mis-using operators in this way.

**Comment by nagisa** [comment]
While it is a bad style to do something like this, it is also perfectly valid code. As rustc’s lints are expected to have no false-positives, this is a genuine bug.

The stylistic preference should be linted by clippy, not rustc.

**Comment by Centril** [comment]
For the record, I disagree that this is bad style, or it's at least debatable. It seems like it's taking advantage of the sugar that `&&` brings over `if`, and Rust being a sideeffectful language, this should be expected.

As for whether this should be linted against or not, it seems to me it should not, but cc @rust-lang/lang if there are objections.

**Comment by eddyb** [comment]
I agree with @nagisa and @Centril, `&&` and `||` should *not* be linted, they're control-flow operators, *not* value operators. Looks like operators like `&` and `|` also warn, so we can keep just those.

**Comment by estebank** [comment]
Pointers:

The code that needs to be modified is

https://github.com/rust-lang/rust/blob/18c275b423f9f13c0e404ae3804967d2ab66337c/src/librustc_lint/unused.rs#L37-L111

I believe you will have to explore the `ExprKind::Binary(_, _, rhs)` recursively to see if the leaf `rhs` is of type `!`, and if that is the case, return early.

**Comment by eddyb** [comment]
> I believe you will have to explore the `ExprKind::Binary(_, _, rhs)` recursively to see if the leaf `rhs` is of type `!`, and if that is the case, return early.

I would just remove this line:
```rust
hir::BinOpKind::And | hir::BinOpKind::Or => Some("logical operation"),
```
As to lint correctly you'd probably also have to look at whether there any side-effects in the RHS.

**Comment by spacekookie** [comment]
@estebank thanks! I'll try to have a look at this and post if I have any questions :)

**Comment by lolbinarycat** [comment]
triage: still an issue on latest nightly

**Comment by TilmanR** [comment]
I'm new to rust and a bit weird when it comes to styling, so I just ran into this issue while trying to make my early returns look distinct from "regular" if statements. (This helps me keep track of guard clauses.)

```Rust
fn example(input: u16) -> u8
{
    input < 37 &&
        return 0;// unused_must_use

    let _ = input > 42 &&
        return 13;// sorry
    
    return 42;
}
```

In such an extreme case the intention of a discarded result seems sufficiently clear. A fix for this would be marvelous.
